File: Cirque/Cirque/Map/Problem/Problem.swift
//
//  Problem.swift
//  Cirque
//
//  Created by Nathan Hadley on 5/18/24.
//

import Turf
import SwiftUI
@_spi(Experimental) import MapboxMaps

@available(iOS 14.0, *)

struct Problem: Identifiable {
    var id = UUID()
    var name: String
    var grade: String
    var order: Int?
    var color: Color
    var description: String?
    var line: [[Int]]
    var topo: String?
    var subarea: String?
    var coordinates: CLLocationCoordinate2D?

    init(problem: Feature) {
        let properties = problem.properties ?? [:]
        name = (properties["name"] as? Turf.JSONValue)?.stringValue ?? "N/A"
        grade = (properties["grade"] as? Turf.JSONValue)?.stringValue ?? "N/A"
        order = getIntFromFeatureProperty(from: properties, forKey: "order")
        
        let colorStr = (properties["color"] as? Turf.JSONValue)?.stringValue
        color = Problem.color(from: colorStr)
        description = (properties["description"] as? Turf.JSONValue)?.stringValue
        
        let lineString = (properties["line"] as? Turf.JSONValue)?.stringValue ?? "[]"
        line = Problem.parseCoordinates(from: lineString)
        
        topo = (properties["topo"] as? Turf.JSONValue)?.stringValue
        subarea = (properties["subarea"] as? Turf.JSONValue)?.stringValue
        
        if let geometry = problem.geometry,
           case let .point(point) = geometry {
            coordinates = point.coordinates
        } else {
            coordinates = nil
        }
    }
    
    static func parseCoordinates(from string: String) -> [[Int]] {
        let data = Data(string.utf8)
        
        do {
            // Parse the JSON data
            if let jsonArray = try JSONSerialization.jsonObject(with: data, options: []) as? [[Int]] {
                return jsonArray
            }
        } catch {
            print("Failed to parse topo line coordinates: \(error.localizedDescription)")
        }
        
        return []
    }
    
    static func color(from colorString: String?) -> Color {
        switch colorString {
        case "blue":
            return Color.blue
        case "white":
            return Color.white
        case "red":
            return Color.red
        case "orange":
            return Color.orange
        case "yellow":
            return Color.yellow
        case "black":
            return Color.black
        default:
            return Color.black
        }
    }
}


extension Turf.JSONValue {
    var stringValue: String? {
        switch self {
        case .string(let value):
            return value
        default:
            return nil
        }
    }
}

File: android/app/src/main/java/com/example/cirque/views/map/problem/Problem.kt
import androidx.compose.ui.graphics.Color
import com.example.cirque.views.map.getInt
import com.example.cirque.views.map.getString
import com.google.gson.JsonObject
import com.mapbox.geojson.Feature
import com.mapbox.geojson.Point
import org.json.JSONArray
import java.util.UUID

data class Problem(
    val id: UUID = UUID.randomUUID(),
    val name: String?,
    val grade: String?,
    val order: Int?,
    val colorStr: String,
    val color: Color,
    val description: String?,
    val line: List<List<Int>>,
    val topo: String?,
    val subarea: String?,
    val coordinates: Point?
) {
    companion object {
        fun fromFeature(feature: Feature): Problem {
            val properties = feature.properties() ?: JsonObject()

            val name = getString(properties, "name")
            val grade = getString(properties, "grade")
            val order = getInt(properties, "order")
            val description = getString(properties, "description")
            val topo = getString(properties, "topo")
            val subarea = getString(properties, "subarea")

            val colorStr = getString(properties, "color") ?: ""
            val color = colorFromString(colorStr)

            val lineString = getString(properties, "line")
            val line = parseTopoLine(lineString)

            val coordinates = (feature.geometry() as? Point)

            return Problem(
                name = name,
                grade = grade,
                order = order,
                colorStr = colorStr,
                color = color,
                description = description,
                line = line,
                topo = topo,
                subarea = subarea,
                coordinates = coordinates
            )
        }

        private fun parseTopoLine(stringCoords: String?): List<List<Int>> {
            if (stringCoords.isNullOrEmpty()) return emptyList()
            return try {
                val jsonArray = JSONArray(stringCoords)
                (0 until jsonArray.length()).map { i ->
                    val innerArray = jsonArray.getJSONArray(i)
                    (0 until innerArray.length()).map { j -> innerArray.getInt(j) }
                }
            } catch (e: Exception) {
                e.printStackTrace()
                emptyList()
            }
        }

        private fun colorFromString(colorString: String?): Color {
            return when (colorString) {
                "blue" -> Color.Blue
                "white" -> Color.White
                "red" -> Color.Red
                "orange" -> Color(0xFFFFA500)
                "yellow" -> Color.Yellow
                "black" -> Color.Black
                else -> Color.Black
            }
        }
    }
}
